scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -0,0 +1,359 @@
+    >>> lap.tosparse()
+    <6x6 sparse array of type '<class 'numpy.int8'>'
+        with 24 stored elements in Compressed Sparse Row format>
+    >>> lap.toarray()
+        array([[-4,  1,  1,  2,  0,  0],
+               [ 1, -4,  1,  0,  2,  0],
+               [ 1,  1, -4,  0,  0,  2],
+               [ 2,  0,  0, -4,  1,  1],
+               [ 0,  2,  0,  1, -4,  1],
+               [ 0,  0,  2,  1,  1, -4]], dtype=int8)
+    >>> np.array_equal(lap.matmat(np.eye(n)), lap.toarray())
+    True
+    >>> np.array_equal(lap.tosparse().toarray(), lap.toarray())
+    True
+    >>> lap.eigs
+    array([-7., -7., -4., -3., -3.,  0.])
+    >>> eigvals = eigvalsh(lap.toarray().astype(np.float64))
+    >>> np.allclose(lap.eigs, eigvals)
+    True
+
+    and with ``'neumann'``
+
+    >>> lap = LaplacianNd(grid_shape, boundary_conditions='neumann')
+    >>> lap.tosparse()
+    <6x6 sparse array of type '<class 'numpy.int8'>'
+        with 20 stored elements in Compressed Sparse Row format>
+    >>> lap.toarray()
+    array([[-2,  1,  0,  1,  0,  0],
+           [ 1, -3,  1,  0,  1,  0],
+           [ 0,  1, -2,  0,  0,  1],
+           [ 1,  0,  0, -2,  1,  0],
+           [ 0,  1,  0,  1, -3,  1],
+           [ 0,  0,  1,  0,  1, -2]])
+    >>> np.array_equal(lap.matmat(np.eye(n)), lap.toarray())
+    True
+    >>> np.array_equal(lap.tosparse().toarray(), lap.toarray())
+    True
+    >>> lap.eigs
+    array([-5., -3., -3., -2., -1.,  0.])
+    >>> eigvals = eigvalsh(lap.toarray().astype(np.float64))
+    >>> np.allclose(lap.eigs, eigvals)
+    True
+
+    """
+
+    def __init__(self, grid_shape, *,
+                 boundary_conditions="neumann",
+                 dtype=np.int8):
+
+        if boundary_conditions not in ("dirichlet", "neumann", "periodic"):
+            raise ValueError(
+                f"Unknown value '{boundary_conditions}' is given for"
+                " 'boundary_conditions' parameter."
+                " The valid options are 'dirichlet', 'periodic', and "
+                "'neumann' (default).")
+
+        self.grid_shape = grid_shape
+        self.boundary_conditions = boundary_conditions
+        self.dtype = dtype
+        # LaplacianNd folds all dimensions in `grid_shape` into a single one
+        N = np.prod(grid_shape)
+        super().__init__(dtype=self.dtype, shape=(N, N))
+
+        indices = np.indices(grid_shape)
+        L = np.zeros(grid_shape)
+
+        for j, n in zip(indices, grid_shape):
+            if boundary_conditions == "dirichlet":
+                L += -4 * np.sin(np.pi * (j + 1) / (2 * (n + 1))) ** 2
+            elif boundary_conditions == "neumann":
+                L += -4 * np.sin(np.pi * j / (2 * n)) ** 2
+            else:  # boundary_conditions == "periodic"
+                L += -4 * np.sin(np.pi * np.floor((j + 1) / 2) / n) ** 2
+
+        self._eigs = np.sort(L.ravel())
+
+    @property
+    def eigs(self):
+        return self._eigs
```

Comment:
I agree. And since we already have a `boundary_conditions` member (also not exactly short), I think we should just use `eigenvalues`.

Reply:
I would prefer it if we didn't move the computation into the property. As a user, if I see something documented as an attribute then I expect it can be accessed with little overhead. However, if we start filling the property with computation, this is not the case. Also, we only need to calculate the eigenvalues once since nothing can change

Reply:
Naming, I really don't have any strong opinion. I can go with the majority vote. For the computation I'm leaning a bit on @j-bowhay 's take on it. It should be indeed independent from the computation. Or at least we can make it lazy method to conpute once but returns the static value.